dex/networks/erc20/contracts/ERC20SwapV0.sol

```diff
@@ -0,0 +1,233 @@
+    // initiating, redeeming, and refunding swaps.
+    constructor() {}
+
+    // senderIsOrigin ensures that this contract cannot be used by other
+    // contracts, which reduces possible attack vectors. There is some
+    // conversation in the eth community about removing tx.origin, which would
+    // make this check impossible.
+    modifier senderIsOrigin() {
+        require(tx.origin == msg.sender, "sender != origin");
+        _;
+    }
+
+    // swap returns a single swap from the swaps map.
+    function swap(bytes32 secretHash)
+        public view returns(Swap memory)
+    {
+        return swaps[secretHash];
+    }
+
+    // Initiation is used to specify the information needed to initiatite a swap.
+    // The ERC20 token address is provided separately to the initiate function.
+    struct Initiation {
+        uint refundTimestamp;
+        bytes32 secretHash;
+        address participant;
+        uint value;
+    }
+
+    // initiate initiates an array of swaps all for the same ERC20 token.
+    // It checks that all of the swaps have a non zero redemptionTimestamp
+    // and value, and that none of the secret hashes have ever been used
+    // previously. Once initiated, each swap's state is set to Filled. The
+    // tokens equal to the sum of each swap's value are now in the custody of
+    // the contract and can only be retrieved through redeem or refund.
+    //
+    // It is important to note that this calls an external contract which has no
+    // restrictions on gas used. This has the potential to open the contract up
+    // to a reentry attack.
+    //
+    // This is a writing function and requires gas. Failure or success should
+    // be guaged by querying the swap and checking state after being mined. Gas
+    // is expended either way.
+    function initiate(Initiation[] calldata initiations, address token)
+        public
+        senderIsOrigin()
+    {
+        uint initVal = 0;
+        for (uint i = 0; i < initiations.length; i++) {
+            Initiation calldata initiation = initiations[i];
+            Swap storage swapToUpdate = swaps[initiation.secretHash];
+
+            require(initiation.value > 0, "0 val");
+            require(initiation.refundTimestamp > 0, "0 refundTimestamp");
+            require(swapToUpdate.state == State.Empty, "dup secret hash");
+
+            swapToUpdate.initBlockNumber = block.number;
+            swapToUpdate.refundBlockTimestamp = initiation.refundTimestamp;
+            swapToUpdate.initiator = msg.sender;
+            swapToUpdate.participant = initiation.participant;
+            swapToUpdate.value = initiation.value;
+            swapToUpdate.state = State.Filled;
+            swapToUpdate.token = token;
+
+            initVal += initiation.value;
+        }
+		bool success;
+		bytes memory data;
+        (success, data) = token.call(abi.encodeWithSelector(TRANSFER_FROM_SELECTOR, msg.sender, address(this), initVal));
```

Comment:
Would the swap contract need to approve itself on the token contract before use?

Reply:
Right, @martonp mentioned this in @JoeGruffins' review here: https://github.com/decred/dcrdex/pull/1360#discussion_r773220069
We'll have to have some code in the client wallet setup or connect methods that ensures it has approved this token swap contract to spend the particular token, and if it has not yet, the user account will have to make an `approve` txn on the token for this swap contract.  :(

Reply:
Ah. I get it now. Thanks.